ext/opentelemetry-ext-jaeger/src/opentelemetry/ext/jaeger/__init__.py

```diff
@@ -0,0 +1,394 @@
+            self.client._seqid = 0
+            #  truncate and reset the position of BytesIO object
+            self.buffer._buffer.truncate(0)
+            self.buffer._buffer.seek(0)
+            self.client.emitBatch(batch)
+            buff = self.buffer.getvalue()
+            if len(buff) > self.max_packet_size:
+                logging.warning(
+                    "Data exceeds the max UDP packet size; size %r, max %r",
+                    len(buff),
+                    self.max_packet_size,
+                )
+            else:
+                udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+                udp_socket.sendto(buff, self.address)
+
+        except Exception as e:  # pragma: NO COVER
+            logging.error(getattr(e, "message", e))
+
+        finally:
+            if udp_socket is not None:
+                udp_socket.close()
+
+
+class Collector:
+    """Submits collected spans to Thrift HTTP server.
+
+    :param thrift_url: URL of the Jaeger HTTP Thrift.
+
+    :param auth: (Optional) Auth tuple that contains username and
+                password for Basic Auth.
+
+    :param http_transport: Class for creating new client for Thrift
+                           HTTP server.
+    """
+
+    def __init__(
+        self,
+        thrift_url: str = "",
+        auth: typing.Tuple[str, str] = None,
+        client=jaeger.Client,
+        http_transport=THttpClient.THttpClient,
+    ):
+        self.thrift_url = thrift_url
+        self.auth = auth
+        self.http_transport = http_transport(uri_or_host=thrift_url)
+        self.client = client(
+            iprot=TBinaryProtocol.TBinaryProtocol(trans=self.http_transport)
+        )
+
+        # set basic auth header
+        if auth is not None:
+            import base64
+
+            auth_header = "{}:{}".format(*auth)
+            decoded = base64.b64encode(auth_header.encode()).decode("ascii")
+            basic_auth = dict(Authorization="Basic {}".format(decoded))
+            self.http_transport.setCustomHeaders(basic_auth)
+
+    def submit(self, batch: jaeger.Batch):
+        """Submits batches to Thrift HTTP Server through Binary Protocol.
+
+        :param batch: Object to emit Jaeger spans.
+        """
+        try:
+            self.client.submitBatches([batch])
+            # it will call http_transport.flush() and
+            # status code and message will be updated
+            code = self.http_transport.code
+            msg = self.http_transport.message
+            if code >= 300 or code < 200:
+                logging.error(
+                    "Traces cannot be uploaded;\
+                        HTTP status code: {}, message {}".format(
+                        code, msg
+                    )
+                )
+        except Exception as e:  # pragma: NO COVER
+            logging.error(getattr(e, "message", e))
+
+        finally:
+            if self.http_transport.isOpen():
+                self.http_transport.close()
```

Comment:
Should we really close http_transport after every submit? Is it automatically reopened? Also, it seems dirty to use both a client that wraps the http transport and the http transport directly, but that might be fine with the thrift library, I know nothing about it.

Reply:
Unfortunately I don't have that many answers for you. It's the way it was done in OpenCensus.
I checked and yes, the transport should be close, it automatically opens when flush is called.